### PR TITLE
completed tut02 review

### PR DESCRIPTION
**Pull request by Edric-Ho**
Enhanced structure and clarity with new resources added. Also based on the course materials provided and considering that there are no prerequisites for this course, only corequisites such as CSC108, CSC110, CSC148, or previous programming experience, I've discussed with Serena. We both agree that courses like CSC108 or CSC110 should be prerequisites for this course, given its coding-oriented approach to introductory statistics. While CSC148 or prior programming experience could be sufficient, having a more robust foundation in programming would enable a smoother and more effective understanding of basic statistics through programming. This also aligns with the GAISE guidelines, which advocate for integrating new technologies and enhancing students' statistical reasoning. This being said solid programming experience should ideally be established before utilizing programming to demonstrate statistical concepts.

Here is a link to a Coursera programming course designed for students without prior programming experience or those looking to refresh their skills before enrolling in CSC148 that might be helpful:  https://www.coursera.org/learn/learn-to-program

### CODE REVIEW

**Review by pointOfive — CHANGES_REQUESTED** [review]
### Requested Changes
- Explore the Monty Hall Problem in detail: [Monty Hall Problem Details](https://chat.openai.com/share/26f74c54-5358-431e-b97a-315fdcb4e2c9)
  - Excellent -- same link as week 1, so we're synched -- great!
  - I think review should be left open and not directed though
  - So make this one of the enumerated points in the "demo" -- I'll let you decide where you think it makes sense to add it into the flow of the demo
- I like the "collapsed" design you've provided -- share this with the group as a design feature they could consider... however... in doing so, please note that
  - this feature does not render on UofT Jupterhub/GitHub, or when "save as .html" is used... 
  - and I'm also not sure that I think the indenting of the collapsing (on colab/vscode?) should include the "Demo [55 minutes]..."... maybe add the collapsing somewhere after the section title so this section is not entirely indented with a collapse button
    - if you do this be careful with the indenting in " 1. `for i in range(n)` , `for x in enumerate(lst)` and `for i,x in enumerate(lst)`\n"... I think an orphaned enumerated list that's indented without a parent item breaks the markdown specification and won't render correctly in several contexts (such as "save as .html or perhaps even google colab... I can't remember exactly on this, but I have experienced the need to be careful with indenting markdown lists) 
    - So what I'm saying here in all of this is "consider bounding the collapse on just the cells that are getting hidden" -- I think that actually works better for using this feature smoothly, if you see what I mean?
- I like your re-organization of the demo outline a lot.
  - I think having a "video" quickly covering these topics as well as a sort of "review" for these ideas after the tutorial (I would think?) is going to give a huge boost opportunity for students... I'm super looking forward to seeing this
  - Once you're recorded that make sure alert the group so this course-build feature/plan/intention gets propagated to / similarly executed in other weeks.
- Code whitespacing/stylization -- I'm very pleased with your changes here -- I guess you're likely following a python style guide -- if so, please forward that to everyone so our python style is uniform across our material.
  - I would like to be more consistent in our spacing usage, e.g., `"# tuple\n"` vs `"#1,2,3\n"`
  - double check where you put your commas -- "secret_winning_door ,so" should be "secret_winning_door, so"
- I can't tell from the GitHub commit diff if you've made any changes to the communication or homework activities
  - That's my own fault for putting everything in a single cell... (which you've reorganized much more effectively by dividing things up)
  - If there are any changes there, can you point me to those in a comment reply here? Then I can just review them in the notebooks directly

**Comment by pointOfive** [comment]
Very interesting comment I am going to keep in mind: "We both agree that courses like CSC108 or CSC110 should be prerequisites for this course, given its coding-oriented approach to introductory statistics."

At the moment we are limited in needing to try to execute my "coding-oriented vision" but it is possible that in the long run it might be decided that I am looking to do "too much" with STA130

**Comment by pointOfive** [comment]
re: "This being said solid programming experience should ideally be established before utilizing programming to demonstrate statistical concepts."

Yep, I hear ya -- so the current intention/plan does not fully respect this (with videos and background material); but, looks to address this issue within the context of the course.

Let's do our best! But continue to keep an eye on our self-evaluation of STA130 with respect to your sentiments.

**Comment by pointOfive** [comment]
I see that https://www.coursera.org/learn/learn-to-program is "enrol for free May 11".

Is this a resource that you think it makes sense for students to utilize within the context of the F24 semester? Will students think "I have to take another course to do this course!!" or can we present a resource like this in a way which feels more like "practice" material associated with the course?

I do appreciate the sentiments you're expressing here, and they may be 100% correct; but, a stats degree is not a CS degree, so there's not time for us to have a CS course as part of our curriculum; while, at the same time, coding is beyond useful for a stats degree, practically speaking... otherwise we just make specialized math students who don't really analyze data in any meaningful real-world sense. 

So, basically, I'm in agreement with you on the challenge here; but, we're in a tricky position of really needing to do a lot to make an effective modern statistician / data scientist. 

I think the correct compromise (for now at least) is that we need to reduce the overall content in the course so that even though we're trying to do "everything", we limit the overall scope enough to make it feasible. Many students experienced the problematic nature of the course exactly in the way you're describing, so the idea here is to add support material (with Chat being a big part of this imo) into the current design of the course to make things a little more helpful and feasible for those students. And even students like yourself and Serena who managed to get by, but can also identify gaps that exist in the course, too.

**Comment by Edric-Ho** [comment]
replies on the original comments:

> ### Requested Changes
> * Explore the Monty Hall Problem in detail: [Monty Hall Problem Details](https://chat.openai.com/share/26f74c54-5358-431e-b97a-315fdcb4e2c9)
>   
>   * Excellent -- same link as week 1, so we're synched -- great!
>   * I think review should be left open and not directed though

Review should be left open and not directed; that's absolutely right. The reason I specifically included the Monty Hall Problem is that students are asked to spend 55 minutes coding it. I believe that if I ask them to review the problem again, they might have a better understanding or recall what the question really is, instead of wasting 55 minutes trying to remember what it is.

>   * So make this one of the enumerated points in the "demo" -- I'll let you decide where you think it makes sense to add it into the flow of the demo
> * I like the "collapsed" design you've provided -- share this with the group as a design feature they could consider... however... in doing so, please note that
>   
>   * this feature does not render on UofT Jupterhub/GitHub, or when "save as .html" is used...
>   * and I'm also not sure that I think the indenting of the collapsing (on colab/vscode?) should include the "Demo [55 minutes]..."... maybe add the collapsing somewhere after the section title so this section is not entirely indented with a collapse button

This issue might be due to not using the UofT Jupyter Hub to review the materials before(DataSpell). There could be some version differences. That's why there are element tags that say 'collapsed': false, but don't contain any actual work even on my own IDE.

>     
>     * if you do this be careful with the indenting in " 1. `for i in range(n)` , `for x in enumerate(lst)` and `for i,x in enumerate(lst)`\n"... I think an orphaned enumerated list that's indented without a parent item breaks the markdown specification and won't render correctly in several contexts (such as "save as .html or perhaps even google colab... I can't remember exactly on this, but I have experienced the need to be careful with indenting markdown lists)
fixed
>     * So what I'm saying here in all of this is "consider bounding the collapse on just the cells that are getting hidden" -- I think that actually works better for using this feature smoothly, if you see what I mean?
> * I like your re-organization of the demo outline a lot.
>   
>   * I think having a "video" quickly covering these topics as well as a sort of "review" for these ideas after the tutorial (I would think?) is going to give a huge boost opportunity for students... I'm super looking forward to seeing this
>   * Once you're recorded that make sure alert the group so this course-build feature/plan/intention gets propagated to / similarly executed in other weeks.

I found a YouTube video that explains the Markov chain, which I think is good. Regarding the pre-class video, here's an update: we are still planning it and awaiting your approval of the course material. Once we have your approval, we can determine what content to include in the short videos. For now, we've made some progress. One or two other teammates have sent me their suggestions/outlines on what they think is most important, and we have started drafting the script.

> * Code whitespacing/stylization -- I'm very pleased with your changes here -- I guess you're likely following a python style guide -- if so, please forward that to everyone so our python style is uniform across our material.
I think what you mean here is the new line spacing? 

So the reason why I added some space/newline regarding the chatbot extension is, I found some lines of code is spacing weirdly what I mean by this is : 

    if words[i+1] in next_word[word]:
        next_word[word][words[i+1]] += 1
    else:

        next_word[word][words[i+1]] = 1
Added a new line right after the else: keyword but did not add a new line after the if condition statement

>   
>   * I would like to be more consistent in our spacing usage, e.g., `"# tuple\n"` vs `"#1,2,3\n"`

fixed!

>   * double check where you put your commas -- "secret_winning_door ,so" should be "secret_winning_door, so"

fixed!

> * I can't tell from the GitHub commit diff if you've made any changes to the communication or homework activities
>   
>   * That's my own fault for putting everything in a single cell... (which you've reorganized much more effectively by dividing things up)
>   * If there are any changes there, can you point me to those in a comment reply here? Then I can just review them in the notebooks directly

Here are the changes I made: 

**Communication section:**
Previous version: 1. [15 minutes] Break into 5 new groups of 5, and assign the following 5 questions to the 5 groups. Consider allowing students to preferentially select which group they join by calling for volunteers for each group.

New:  1.[15 minutes] Break into 5 new groups of 4-5, and assign the following 5 questions to the 5 groups. Consider allowing students to preferentially select which group they join by calling for volunteers for each group. They should consider the pros and cons of each option, the potential impact of their decision, and how uncertainties affect outcomes.

previous version:
2. Healthcare Treatment Decision: Students are healthcare professionals deciding between two treatment options for a patient's condition. Each treatment has different success rates and potential side effects.
        1. Treatment A: High success rate but moderate side effects.
 (not indented)2. Treatment B: Lower success rate but minimal side effects.


new: 2. Healthcare Treatment Decision: Students are healthcare professionals deciding between two treatment options for a patient's condition. Each treatment has different success rates and potential side effects.
        1. Treatment A: High success rate but moderate side effects.
        (indented)2. Treatment B: Lower success rate but minimal side effects.'

new points added at the end of this section:
2. **[20 - 25 minutes]** Each group prepares a brief summary, approximately 3 minutes, outlining their decision and the rationale behind it. They should address the expected outcomes, the risks involved, and why they believe their choice is the best given the uncertainty. After each presentation, allow a 1-2 minutes Q&A session where other students or TAs can ask questions or challenge the decision.

**Homework section:**

Instructions with new points added:
3. Attempt to code (WITH COMMENTS, TO HELP BETTER COMMUNICATION) the Monty Hall problem on your own, without using external aids. You can either develop your own method or use the approach originally introduced by Prof. Scott to become more familiar with basic operations and Python syntax. This exercise will also help you gain a better understanding of functions in the NumPy library.
4.  Watch the embedded video tutorial on Markov chains in the next Jupyter cell to understand their application in chatbot creation. After viewing, tell ChatGPT that you have code that creates a Markov chatbot and show it the code below and then use ChatGPT to make sure you understand how the code works and the problem works.

New code added:
#Load dataset
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-08-11/avatar.csv"
avatar = pd.read_csv(url)
words = ("\n"+avatar.character.str.upper().str.replace(' ','.')+": "+avatar.full_text+" ").sum().split(' ')

File: TUT/STA130F24_TUT02_Sep13.ipynb
Comment (pointOfive) [inline TUT/STA130F24_TUT02_Sep13.ipynb:8]: Yep for now -- discussed in comment point 1.

File: TUT/STA130F24_TUT02_Sep13.ipynb
Comment (pointOfive) [inline TUT/STA130F24_TUT02_Sep13.ipynb:25]: Good -- yep this better is outlining -- I'm going to add one more level to outline and put "Review Monte Hall" here per my point 1 comment already noted; but, you'll hopefully be pleased to note that I use the word "Review" (just need to place it in different section)

File: TUT/STA130F24_TUT02_Sep13.ipynb
Comment (pointOfive) [inline TUT/STA130F24_TUT02_Sep13.ipynb:51]: hmm weird, seems out of place? There's not code for it? I'll clean it up later if needed.

File: TUT/STA130F24_TUT02_Sep13.ipynb
Comment (pointOfive) [inline TUT/STA130F24_TUT02_Sep13.ipynb:86]: yeah these all seemed good at a quick glance -- thanks!

File: TUT/STA130F24_TUT02_Sep13.ipynb
Comment (pointOfive) [inline TUT/STA130F24_TUT02_Sep13.ipynb:128]: yeah good this is just moved from above with perhaps some small edits (enumerate 4/4 fixed to 4/5, e.g.)

Comment (pointOfive) [reply TUT/STA130F24_TUT02_Sep13.ipynb:128]: Yep -- changes noted by Edric -- I'll have one more double check later but adjustments seem like improvements

File: TUT/STA130F24_TUT02_Sep13.ipynb
Comment (pointOfive) [inline TUT/STA130F24_TUT02_Sep13.ipynb:129]: Okay, we have an issue here of my own making: we have only 100 minutes of time to use, but now have
- review: 20
- demo: 55
- comm: 15
- speech: 20-25

My current thinking is to adjust this to:
- review: 15
- demo: 50 (less time than I promised Edric in my comment above)
- comm: 15
- speech: 20

Comment (pointOfive) [reply TUT/STA130F24_TUT02_Sep13.ipynb:129]: Also. Edric added some further clarification instruction here which looks like a positive addition

File: TUT/STA130F24_TUT02_Sep13.ipynb
Comment (pointOfive) [inline TUT/STA130F24_TUT02_Sep13.ipynb:139]: Moved from above.
Edric noted (just like for above) his additions in his response comment.

File: TUT/STA130F24_TUT02_Sep13.ipynb
Comment (pointOfive) [inline TUT/STA130F24_TUT02_Sep13.ipynb:213]: Nice (all of these)

File: TUT/STA130F24_TUT02_Sep13.ipynb
Comment (pointOfive) [inline TUT/STA130F24_TUT02_Sep13.ipynb:263]: oh and a few more. Great.

**Review by pointOfive — APPROVED** [review]
Nice -- as far as I can tell you've addressed everything very comprehensively.

1. Monte Hall review placement quibbles: haha no you're gaslighting 55 minutes of not understanding. Definitely reviewing it is a good idea, but the question is how much time we allocate to free review versus doing the Monte Hall question, which should be reviewed as part of starting up the Monte Hall coding.  
    1. I think your strong preference here could be based on (a) us splitting hairs on the "proper" usage of the term "review" or (b) you thinking 55 minutes of coding Monte Hall is not enough; but, I think the TAs can be expected to be prepared to code this up efficiently. I'm expecting to give them 1.5 hours of preparation. I don't want to skip on legit review questions of other stuff for 20 minutes... I'll drop it to 15 minutes as a compromise; but, I don't want to signal that review should be about the topic that's going to be discussed -- it should be for free and open review driven by student questions. I'll make this change after since I feel *so strongly* about it ;) :P
2. I'm not clear here: are you saying you're not actually trying to utilize collapsing cells? And that just happened through the IDE you were using? If that's the case I'll just upload an uncollapsed version of the notebook so there's no extraneous "data" in the notebook.
3. Oo Markov video cool -- lemme see that (perhaps later in your updates) -- this would need to be "simple" but I think the idea that the next word depends on the previous is, or that it depends on the previous words is, etc.; although, I did get some concern about this from the other profs so I'm keeping this topic "under advisement" at the moment.
    1. awesome on the videos, great to hear. Yeah, for me the topics here that are relevant are `for` loops and `if` stuff... `tuple`/`dict` that sort of stuff is what I had in mind... probably/hopefully what you guys are thinking/seeing as well.
4. spacing edits -- good -- thanks for keeping on eye on this as you're able! And, oo yeah ick, line break inconsistency for `if/else` def appreciate you correcting that 
5. Changes.  Thanks for listing these for me to glance over -- much appreciated, that was very easy to look at!
    1. I was pleased with everything here.
    2. I might tinker with language a bit, but I just always tend to do that (even with my own stuff all the time). I thought the added ideas were positive additions.  

#### Ugh... I can't add to 100 -- I'm gonna have to look at this again as we need the minutes to sum to 100 to represent starting 10 minutes after the hour and having a 10 minute break again (so 120 minutes - 20)

And I'm pulling in (fresh) week 9 and I'll look at it from there!

**Pull request merged**
